x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitForRolloverReadyStep.java

```diff
@@ -228,98 +227,50 @@ public void evaluateCondition(Metadata metadata, Index index, Listener listener,
     RolloverRequest createRolloverRequest(String rolloverTarget, TimeValue masterTimeout, boolean rolloverOnlyIfHasDocuments) {
         RolloverRequest rolloverRequest = new RolloverRequest(rolloverTarget, null).masterNodeTimeout(masterTimeout);
```

Comment:
Could the `RolloverRequest` benefit from knowing about the new `RolloverConditions` entity?

Reply:
After some further investigation, it's not that simple, this is the outcome:
- I decided to use the Condition name in the `RolloverConditions`, this way we are guaranteed that names are the same and there is no chance of having a typo in `RolloverConditions`.
- The `RolloverRequest` is using a Map of conditions instead of a dedicated object. This makes the parsing of the input a bit harder but it makes the `areConditionsMet` very readable:
```
public boolean areConditionsMet(Map<String, Boolean> conditionResults) {
        boolean allMinConditionsMet = conditions.values()
            .stream()
            .filter(c -> Condition.Type.MIN == c.type())
            .allMatch(c -> conditionResults.getOrDefault(c.toString(), false));

        boolean anyMaxConditionsMet = conditions.values()
            .stream()
            .filter(c -> Condition.Type.MAX == c.type())
            .anyMatch(c -> conditionResults.getOrDefault(c.toString(), false));

        return conditionResults.size() == 0 || (allMinConditionsMet && anyMaxConditionsMet);
    }
```
I do not think there is a benefit to change this. What I am considering though is that we could use the map of conditions for both `ILM` & `DLM`. I will double check this.

Reply:
So, the PR grew a bit more than I expected but I managed to collect all the rollover condition logic in one class `RolloverConditions`. It uses a map of conditions as a model, which is what the `RolloverRequest` uses. This means that now ILM, DLM and the `RolloverRequest` all use the same model and we do not have to convert one object to the other. Furthermore, all the conditions are defined in one place and shared everywhere.

The `RolloverAction` still requires special attention with wire serialization to be backwards compatible.

I noticed that ILM and the rollover request have different validations:
- ILM requires at least one `max*` condition
- `RolloverRequest` requires at least one `max*` condition but only if there is at least one `min*` condition, that's why for now the validation is not part of the `RolloverConditions`.

Furthermore, I introduced a builder for the `RolloverConditions`. I am not 100% sure about this, I understand it's extra object creation, but I thought it was important that the `RolloverConditions` are immutable now that we pass them around more. I am open to better ideas :).

Since this code changed significantly since the last review, I would like to ask @andreidan & @dakrone you feedback for one more time.

Reply:
Thanks Mary, this LGTM. I think it makes sense now that the `RolloverRequest` and `TransportRolloverAction` also know about `RolloverConditions` 🚀 

I like it ! I also like the `Builder`and don't think it's necessarily excessive to have an extra allocation.

The validations for the rollover conditions are ... a bit ad-hoc at the moment (ie. we should also do much more like for example validating that using max_size and max_primary_shard_size makes sense - former > latter etc) and we should consolidate them at some point, thanks for keeping them "as-is" in this PR.

Reply:
I am glad you agree with the result, it felt really nice working on this! 🚀